Changes in module 'nf-core/instrain/compare'
--- /dev/null
+++ modules/nf-core/instrain/compare/nextflow.config
@@ -0,0 +1,19 @@
+process {
+    withName: INSTRAIN_COMPARE {
+        publishDir = [
+            [
+                path: { "${params.outdir}/VirusMicrodiversity/instrain/compare" },
+                mode: params.publish_dir_mode,
+                pattern: '*.IS_compare',
+            ]
+        ]
+        ext.args   = [
+            params.instrain_min_variant_cov ? "--min_cov ${params.instrain_min_variant_cov}" : "",
+            params.instrain_min_snp_freq ? "--min_freq ${params.instrain_min_snp_freq}" : "",
+            params.instrain_max_snp_fdr ? "--fdr ${params.instrain_max_snp_fdr}" : "",
+            params.instrain_min_genome_breadth ? "--breadth ${params.instrain_min_genome_breadth}" : "",
+            params.instrain_popani_thresh ? "--ani_threshold ${params.instrain_popani_thresh}" : "",
+            params.instrain_min_genome_comp ? "--coverage_threshold ${params.instrain_min_genome_comp}" : "",
+        ].join(' ').trim()
+    }
+}
--- modules/nf-core/instrain/compare/main.nf
+++ modules/nf-core/instrain/compare/main.nf
@@ -8,13 +8,16 @@
         'biocontainers/instrain:1.6.1--pyhdfd78af_0' }"
 
     input:
-    tuple val(meta), path(profiles)
-    tuple val(meta2), path(bams)
+    tuple val(meta), path(bams), path(profiles)
     path stb_file
 
     output:
-    tuple val(meta), path("*.IS_compare")   , emit: compare
-    path "versions.yml"                     , emit: versions
+    tuple val(meta), path("*.IS_compare")                                               , emit: compare
+    tuple val(meta), path("*.IS_compare/output/*.IS_compare_comparisonsTable.tsv")      , emit: comparisons_table
+    tuple val(meta), path("*.IS_compare/output/*.IS_compare_pooled_SNV_data.tsv")       , emit: pooled_snv
+    tuple val(meta), path("*.IS_compare/output/*.IS_compare_pooled_SNV_data_keys.tsv")  , emit: snv_keys
+    tuple val(meta), path("*.IS_compare/output/*.IS_compare_pooled_SNV_info.tsv")       , emit: snv_info
+    path "versions.yml"                                                                 , emit: versions
 
     when:
     task.ext.when == null || task.ext.when

************************************************************
