Changes in module 'nf-core/mash/sketch'
--- modules/nf-core/mash/sketch/main.nf
+++ modules/nf-core/mash/sketch/main.nf
@@ -24,9 +24,9 @@
     mash \\
         sketch \\
         $args \\
+        $reads \\
         -p $task.cpus \\
         -o ${prefix} \\
-        -r $reads \\
         2> ${prefix}.mash_stats
 
     cat <<-END_VERSIONS > versions.yml
@@ -34,4 +34,17 @@
         mash: \$(mash --version 2>&1)
     END_VERSIONS
     """
+
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch ${prefix}.msh
+    touch ${prefix}.mash_stats
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        mash: \$(mash --version 2>&1)
+    END_VERSIONS
+    """
 }

************************************************************
