Changes in module 'nf-core/custom/dumpsoftwareversions'
--- /dev/null
+++ modules/nf-core/custom/dumpsoftwareversions/nextflow.config
@@ -0,0 +1,9 @@
+process {
+    withName: CUSTOM_DUMPSOFTWAREVERSIONS {
+        publishDir = [
+            path: { "${params.outdir}/pipeline_info" },
+            mode: params.publish_dir_mode,
+            pattern: '*_versions.yml'
+        ]
+    }
+}
--- modules/nf-core/custom/dumpsoftwareversions/main.nf
+++ modules/nf-core/custom/dumpsoftwareversions/main.nf
@@ -21,4 +21,13 @@
     script:
     def args = task.ext.args ?: ''
     template 'dumpsoftwareversions.py'
+
+    stub:
+    def args = task.ext.args ?: ''
+    template 'dumpsoftwareversions.py'
+    """
+    touch software_versions.yml
+    touch software_versions_mqc.yml
+    touch versions.yml
+    """
 }

************************************************************
