Changes in module 'nf-core/checkv/endtoend'
--- modules/nf-core/checkv/endtoend/main.nf
+++ modules/nf-core/checkv/endtoend/main.nf
@@ -16,8 +16,8 @@
     tuple val(meta), path ("${prefix}/completeness.tsv")    , emit: completeness
     tuple val(meta), path ("${prefix}/contamination.tsv")   , emit: contamination
     tuple val(meta), path ("${prefix}/complete_genomes.tsv"), emit: complete_genomes
-    tuple val(meta), path ("${prefix}/proviruses.fna")      , emit: proviruses
-    tuple val(meta), path ("${prefix}/viruses.fna")         , emit: viruses
+    tuple val(meta), path ("${prefix}/proviruses.fna.gz")   , emit: proviruses
+    tuple val(meta), path ("${prefix}/viruses.fna.gz")      , emit: viruses
     path "versions.yml"                                     , emit: versions
 
     when:
@@ -35,6 +35,8 @@
         -d $db \\
         $fasta \\
         $prefix
+
+    gzip ${prefix}/*viruses.fna
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
@@ -55,6 +57,8 @@
     touch ${prefix}/proviruses.fna
     touch ${prefix}/viruses.fna
 
+    gzip ${prefix}/*viruses.fna
+
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         checkv: \$(checkv -h 2>&1  | sed -n 's/^.*CheckV v//; s/: assessing.*//; 1p')

--- /dev/null
+++ modules/nf-core/checkv/endtoend/nextflow.config
@@ -0,0 +1,16 @@
+process {
+    withName: CHECKV_ENDTOEND {
+        publishDir = [
+            [
+                path: { "${params.outdir}/VirusQuality/checkv/endtoend" },
+                mode: params.publish_dir_mode,
+                pattern: '**/quality_summary.tsv'
+            ],
+            [
+                path: { "${params.outdir}/VirusQuality/checkv/endtoend" },
+                mode: params.publish_dir_mode,
+                pattern: '**/*viruses.fna.gz'
+            ]
+        ]
+    }
+}
************************************************************
